### ISSUE TITLE
Errors when using 'exact' representation of density matrices

### ISSUE THREAD

**Issue by PhilipVinc** [netket/netket#607]
A worrisome email I received:

>Since the 3.0 update some observables in my calculations switched signs (the MC estimates).
>
>When calculating expectation values of an operator that is not symmetric in a steady state solution I get a different sign when calculating
>```
>rho = nk.exact.steady_state(lind)
>tr(rho@op.to_dense())
>0.15
>ss.estimate(op)
>-0.15
>```
>
>When investigating this, I didn’t find any fault in the MC nk code, but one possible explanation could be that operator.to_dense() returns a transposed matrix or is this your convention that operators are applied to the left?

He also showed me a case that I further reduced a bit. here is my explanation:

In nette, for reasons unbeknown to me, we have the convention of labelling the Hilbert space from down to up, from low to high. So the local Hilbert space of Pauli matrices is [down, up], as testified by
```
>>> nk.hilbert.Spin(0.5).all_states()
array([[-1.],
       [ 1.]])
 
```

However, if I now create the sigma plus operator I get:
```
>>> nk.operator.spin.sigmap(nk.hilbert.Spin(0.5), 0).operators
[array([[0., 1.],
       [0., 0.]])]
```

And clearly in this conventiont this is the sigma_minus operator.
The same issue happens with non symmetric matrices like sigma_y.

The error is on my part, as I had taken that code from some older code of mine which I guess assumed opposite ordering. Interestingly enough, boson operators are correct...

This is consistent with the fact that I long have seen that sigma_y had the opposite sign in some calculations compared to my old article, but after investigating I thought the error was on my old article...

Still, while the fix is clear, I have to double check the Impact on the ordering of the lindblad jump operators because I think I might have implemented a transposition there otherwise I don't understand how it can work...

---

Note: THIS WILL BREAK CODES

**Comment by PhilipVinc** [comment]
I think also sigma_z is wrong.

**Comment by PhilipVinc** [comment]
Also Ising and Heisenberg are wrong.
They use the transposed sigma_y, but the wrong sigma_z.

So they behave fine until you have to measure some observable like sigma_plus/minus on them.

The culprit seems to be the definition that NetKet has used since the dawn of times,
```python
sx = np.array([[0, 1], [1, 0]]);
sy = np.array([[0, -1j], [1j, 0]]);
sz = np.array([[1, 0], [0, -1]]); (edited) 
```

is... wrong, as it does not respect the Pauli algebra:

From Pauli algebra we know that Sx@Sy = 1j Sz

try multiplying sx@sy above, it does not give the correct result.

**Comment by PhilipVinc** [comment]
Mmmm.. 
no, it seems that I am exaggerating.
The historical definition of the Pauli matrices is correct, and the only bug is me transposing them in operator.spin.

It's simply that we are using an opposite convention for the states, where [-1] corresponds to the up state which is... weird.

**Comment by gcarleo** [comment]
Exactly, we are using a computational basis which is not corresponding to the eigenstates of sigma_z, but rather -sigma_z. This is maybe confusing, but I think that defining raising and lowering operators consistently will not lead to wrong results...

**Comment by VolodyaCO** [comment]
I noticed the issue a [while ago](https://github.com/netket/netket/issues/469#issuecomment-692725941), so in my codes I just flip the states at the end if I want to compute anything out of the states. This, indeed, only happens with spins and not bosons as the +1 state normally corresponds to (1,0) and the -1 state normally corresponds to (0,1), but in netket it's the other way around.

Also, by the definitions you give (@PhilipVinc) of the pauli matrices, they are correct and the commutation relations are correct. Maybe a solution would be to leave things as they are, but when someone wants to sample the steady state or the ground state, or create a batch of samples from a system with spins, there should be a warning explaining the issue (maybe some future warning saying that in a future version of netket things will be flipped for spins) and some util, to help the user flip things around for the spin part of the system.

**Comment by gcarleo** [comment]
A solution might be to just redefine ```Spin``` in a way that is sorted the other way around... it's not a big issue I think. The important is just that to avoid bugs ```sigmap``` and ```sigmam``` are correctly and consistently defined

**Comment by jmellak** [comment]
Hi, I have brought up the initial issue with the expectation values. 
A redefinition of the sigma_plus and minus changes the signs of both calculations, so maybe there is a mixing of conventions? I don't use sigma_x,y,z. 

`rho = nk.exact.steady_state(lind)
tr(rho@op.to_dense())
-0.15
ss.estimate(op)
0.15`

tr(rho@op.to_dense().T) gives the same result, but this changes things...

**Comment by PhilipVinc** [comment]
What is op?

**Comment by jmellak** [comment]
Unfortunate naming on my part - op is an operator representing an observable (an asymmetric one, e.g. Sigma_y).

I attached a minimal demo script based on the dissipative ising1d example, showing what exactly I mean.
[testOperators.py.zip](https://github.com/netket/netket/files/6288081/testOperators.py.zip)
The output I get:  
[testOperators_output.txt](https://github.com/netket/netket/files/6288107/testOperators_output.txt)
Can you reproduce this? Takes a minute to run. 

Here 
```
print(f'RBM result for the observable: {ss.estimate(obs)}')
print(f'Exact result from ED:          {np.trace(rho_0@obs.to_dense())}')
```
Gives the output:
> ...
> RBM result for the observable: 0.063+0.021j ± 0.052 [σ²=1.360, R̂=0.9978]
> Exact result from ED:          (-0.08141487086313647+1.5265566588595902e-16j)
> ...

In the second half of the script I recreated the netket mechanism for calculating rho.expect(operator) to better understand what is going on. 
Using this code and removing the Monte-Carlo part I get the exact same result as the ED result rho_0 but with a different sign.

> ...
> Reproduction of expectation value using rho_0: (0.08141487086313648-1.1102230246251565e-16j)
> Result from exact diagonalization rho_0      : (-0.08141487086313647+1.5265566588595902e-16j)
> ...

But if I transpose all operator matrices, which would change the sign of the observable, the lindblad operator is not correct. 
I hope you can reproduce my results with this and maybe tell me where I messed up?

**Comment by gcarleo** [comment]
I think that the "fix" to this is to re-order ```Spin``` basis states in order to have [1,-1] instead of [-1,1], this is a one-line fix

**Comment by gcarleo** [comment]
But again, I stress that this is not a bug, one can well call the basis states [-3435,+231212], define the usual Pauli matrices, and still have consistent results. (modulo knowing that the quantum numbers used are not the eigenstates of Sigma_z). The issue I think here arises only in the density matrix, to be investigated further ...

**Comment by PhilipVinc** [comment]
Upon more debugging it seems to me that:
 - `nk.exact.steady_state` returns the transpose of the steady_state. I say that by comparing the steady state computed by it with the steady state given by qutip.

**Comment by PhilipVinc** [comment]
Upon more debugging it seems to me that:

 - `nk.exact.steady_state` returns the transpose of the steady_state. I say that by comparing the steady state computed by it with the steady state given by qutip. (I'm sure of that)
 - `nk.MCMixedState.to_matrix()` returns the transpose of the state (I'm 95% sure of this)
 - `nk.MCMixedState.estimate(obs)` works correctly

The fix would be fixing the two methods above...

**Comment by jmellak** [comment]
Thank you for digging into this! 👍 

I just want to mention that the 3.0 version release changed the MC result compared to the development branch v3.0, the exact result stayed the same. Maybe this helps pinpointing.

**Comment by PhilipVinc** [comment]
Upon further inspection, we store operators and super-operators with the same ordering of qutip.
So it was baffling to see that when I look for the steady state of a liouvillian by exact diagonalisation we get the transpose.

But then I found [qutip.superoperator.vec2mat](https://github.com/qutip/qutip/blob/67c46870749b6e5925feed106a493eac9124bcc3/qutip/superoperator.py#L322), and it seems that qutip, every time it reshapes from vector to matrix and viceversa it transposes the matrix.

Which we are not doing. So this explains it... 

I remember investigating a lot this ordering (it boils down to the way you flatten the density matrix) and I was sure I had found the correct way to do this in, but I have to re-ivnestigate this now...

Still I wonder if this affects the Monte Carlo code
